src/tools/ceph_dedup_tool.cc

```diff
@@ -525,6 +534,369 @@ void ChunkScrub::chunk_scrub_common()
+
+  IoCtx io_ctx;
+  IoCtx chunk_io_ctx;
+  std::list<chunk_t> duplicable_chunks;
+  size_t total_duplicated_size = 0;
+  size_t total_object_size = 0;
+
+  std::set<std::string> oid_for_evict;
+  size_t chunk_size = 0;
+  pg_pool_t::fingerprint_t fp_type = pg_pool_t::TYPE_FINGERPRINT_NONE;
+  std::string chunk_algo;
+  SampleDedupGlobal &sample_dedup_global;
+  ObjectCursor begin;
+  ObjectCursor end;
+};
+
+void SampleDedupWorkerThread::crawl()
+{
+  cout << "new iteration" << std::endl;
+
+  for (ObjectCursor current_object = begin;
+      current_object < end;) {
+    std::vector<ObjectItem> objects;
+    // Get the list of object IDs to deduplicate
+    std::tie(objects, current_object) = get_objects(current_object, end, 100);
+
+    // Pick few objects to be processed. Sampling ratio decides how many
+    // objects to pick. Lower sampling ratio makes crawler have lower crawling
+    // overhead but find less duplication.
+    auto sampled_indexes = sample_object(objects.size());
+    for (size_t index : sampled_indexes) {
+      ObjectItem target = objects[index];
+      try_dedup_and_accumulate_result(target);
+    }
+  }
+
+  vector<AioCompRef> evict_completions(oid_for_evict.size());
+  int i = 0;
+  for (auto &oid : oid_for_evict) {
+    auto completion = do_async_evict(oid);
+    evict_completions[i] = move(completion);
+    i++;
+  }
+  for (auto &completion : evict_completions) {
+    completion->wait_for_complete();
+  }
+  cout << "done iteration" << std::endl;
+}
+
+AioCompRef SampleDedupWorkerThread::do_async_evict(string oid)
+{
+  Rados rados;
+  ObjectReadOperation op_tier;
+  AioCompRef completion(rados.aio_create_completion());
+  op_tier.tier_evict();
+  io_ctx.aio_operate(
+      oid,
+      completion.get(),
+      &op_tier,
+      NULL);
+  return completion;
+}
+
+std::tuple<std::vector<ObjectItem>, ObjectCursor> SampleDedupWorkerThread::get_objects(
+  ObjectCursor current, ObjectCursor end, size_t max_object_count)
+{
+  std::vector<ObjectItem> objects;
+  ObjectCursor next;
+  int ret = io_ctx.object_list(
+    current,
+    end,
+    max_object_count,
+    {},
+    &objects,
+    &next);
+  if (ret < 0 ) {
+    cerr << "error object_list" << std::endl;
+    objects.resize(0);
+  }
+
+  return std::make_tuple(objects, next);
+}
+
+std::vector<size_t> SampleDedupWorkerThread::sample_object(size_t count)
+{
+  std::vector<size_t> indexes(count);
+  for (size_t i = 0 ; i < count ; i++) {
+    indexes[i] = i;
+  }
+  default_random_engine generator;
+  shuffle(indexes.begin(), indexes.end(), generator);
+  size_t sampling_count = static_cast<double>(count) *
+    sample_dedup_global.sampling_ratio;
+  indexes.resize(sampling_count);
+
+  return indexes;
+}
+
+void SampleDedupWorkerThread::try_dedup_and_accumulate_result(ObjectItem &object)
+{
+  bufferlist data = read_object(object);
+  if (data.length() == 0) {
+    cerr << __func__ << " skip object " << object.oid
+      << " dedup (read failed)\n";
```

Comment:
What if the object were actually length 0?  Message should simply indicate that the object has length 0.  Use std::endl rather than \n.